### PR TITLE
Em/feat/muinversion

### PR DESCRIPTION
**Pull request by lheagy**
Machinery to invert for magnetic permeability from FDEM data
- derivatives and tests for all 4 FDEM formulations
- example of inverting for mu_r (relative permeability)
- addition of a Mu_relative map, ChiMap
- Cyl-mesh specific interpolation (important for CCV interpolation) 

image

@rowanc1 : could you take a look?

### CODE REVIEW

File: SimPEG/EM/Base.py [hunk 1]
```diff
@@ -126,6 +126,19 @@ def MfMui(self):
             self._MfMui = self.mesh.getFaceInnerProduct(self.mui)
         return self._MfMui
 
+    def MfMuiDeriv(self, u):
+        """
+        Derivative of :code:`MfMui` with respect to the model.
+        """
+        if (
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: minor comment on style:

I would return Zero early. It is the easy case to exclude and allows you to whittle down assumptions and get to the hard things last.

File: SimPEG/EM/Base.py [hunk 2]
```diff
@@ -135,6 +148,23 @@ def MfMuiI(self):
             self._MfMuiI = self.mesh.getFaceInnerProduct(self.mui, invMat=True)
         return self._MfMuiI
 
+    # TODO: This should take a vector
+    def MfMuiIDeriv(self, u):
+        """
+        Derivative of :code:`MeSigma` with respect to the model
+        """
+        # TODO: only works for diagonal tensors. getEdgeInnerProductDeriv,
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: can we raise an exception instead? There are tensor utils available to figure this out.

File: SimPEG/EM/Base.py [hunk 3]
```diff
@@ -135,6 +148,23 @@ def MfMuiI(self):
             self._MfMuiI = self.mesh.getFaceInnerProduct(self.mui, invMat=True)
         return self._MfMuiI
 
+    # TODO: This should take a vector
+    def MfMuiIDeriv(self, u):
+        """
+        Derivative of :code:`MeSigma` with respect to the model
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: Incorrect doc string.

File: SimPEG/EM/Base.py [hunk 4]
```diff
@@ -145,6 +175,19 @@ def MeMu(self):
             self._MeMu = self.mesh.getEdgeInnerProduct(self.mu)
         return self._MeMu
 
+    def MeMuDeriv(self, u):
+        """
+        Derivative of :code:`MfMui` with respect to the model.
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: `MfMu` not `MfMui`

File: SimPEG/EM/Base.py [hunk 5]
```diff
@@ -154,6 +197,23 @@ def MeMuI(self):
             self._MeMuI = self.mesh.getEdgeInnerProduct(self.mu, invMat=True)
         return self._MeMuI
 
+    # TODO: This should take a vector
+    def MeMuIDeriv(self, u):
+        """
+        Derivative of :code:`MeSigma` with respect to the model
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: doc string.

File: SimPEG/EM/Base.py [hunk 6]
```diff
@@ -154,6 +197,23 @@ def MeMuI(self):
...
+        """
+        # TODO: only works for diagonal tensors. getEdgeInnerProductDeriv,
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: I would again raise a `NotImplementedError`

File: SimPEG/EM/Base.py [hunk 7]
```diff
@@ -154,6 +197,23 @@ def MeMuI(self):
...
+        #       invMat=True should be implemented in SimPEG
+
+        if (
+            getattr(self, 'muMap', None) is not None or
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: Do we need to be careful with the attr not existing?
Do we have to check both mu and mui maps? If one is set the other should be too?

```python
if self.muMap is None:
    return Utils.Zero()
dMeMuI_dI = -self.MeMuI**2
dMe_dmu = self.mesh.getEdgeInnerProductDeriv(self.mu)(u)
return dMeMuI_dI * (dMe_dmu * self.muDeriv)
```

I think that is much cleaner. Would need to validate that `muiMap` not being `None` would still work.

File: SimPEG/EM/Base.py [hunk 8]
```diff
@@ -170,7 +230,15 @@ def MeSigmaDeriv(self, u):
         """
         Derivative of MeSigma with respect to the model
         """
-        return self.mesh.getEdgeInnerProductDeriv(self.sigma)(u) * self.sigmaDeriv
+        if (
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: same comment here. I think we can just use `sigmaMap`?

File: SimPEG/EM/Base.py [hunk 9]
```diff
@@ -189,9 +257,14 @@ def MeSigmaIDeriv(self, u):
         # TODO: only works for diagonal tensors. getEdgeInnerProductDeriv,
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: raise exception?

File: SimPEG/EM/FDEM/FieldsFDEM.py [hunk 10]
```diff
@@ -344,11 +345,12 @@ def startup(self):
         self._MeSigma = self.survey.prob.MeSigma
         self._MeSigmaDeriv = self.survey.prob.MeSigmaDeriv
         self._MfMui = self.survey.prob.MfMui
+        self._MfMuiDeriv = self.survey.prob.MfMuiDeriv
 
     def _GLoc(self, fieldType):
         if fieldType == 'e':
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/FieldsFDEM.py]:
```python
in ['e', 'eSecondary']
```

??

File: SimPEG/EM/FDEM/ProblemFDEM.py [hunk 11]
```diff
@@ -361,20 +441,41 @@ def getADeriv(self, freq, u, v, adjoint=False):
...
+            return MfMuiDeriv.T * (C * (MeSigmaI.T * (C.T * v)))
+        return C * (MeSigmaI * (C.T * (MfMuiDeriv * v)))
+
+    def getADeriv(self, freq, u, v, adjoint=False):
+        if adjoint is True and self._makeASymmetric:
+            v = self.MfMui * v
+
+        ADeriv =  (
+            self.getADeriv_sigma(freq, u, v, adjoint) +
+            self.getADeriv_mui(freq, u, v, adjoint)
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: ha. so easy.

File: SimPEG/EM/FDEM/ProblemFDEM.py [hunk 12]
```diff
@@ -361,20 +441,41 @@ def getADeriv(self, freq, u, v, adjoint=False):
...
+        MfMuiDeriv = self.MfMuiDeriv(u)
+        MeSigmaI = self.MeSigmaI
+        C = self.mesh.edgeCurl
+
+        if adjoint:
+            return MfMuiDeriv.T * (C * (MeSigmaI.T * (C.T * v)))
+        return C * (MeSigmaI * (C.T * (MfMuiDeriv * v)))
+
+    def getADeriv(self, freq, u, v, adjoint=False):
+        if adjoint is True and self._makeASymmetric:
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: perhaps we need some more validation thoughts.

If all properties are not invertible. Etc. This should be introduced as a slightly more general concept. Could you help me think through some ideas of things to check:

- [ ] has a model (to do an early exit of zero)
- [ ] all physical properties are satisfied or have mappings
- [ ] the mappings are of the right size

The model being set should do some of these checks as well. and the `m=None` on fields and data calculation.

File: SimPEG/EM/FDEM/RxFDEM.py [hunk 13]
```diff
@@ -11,17 +12,27 @@ class BaseRx(SimPEG.Survey.BaseRx):
     """
 
     def __init__(self, locs, orientation=None, component=None):
-        assert(orientation in ['x','y','z']), "Orientation {0!s} not known. Orientation must be in 'x', 'y', 'z'. Arbitrary orientations have not yet been implemented.".format(orientation)
-        assert(component in ['real', 'imag']), "'component' must be 'real' or 'imag', not {0!s}".format(component)
+        assert(
+            orientation in ['x', 'y', 'z']
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/RxFDEM.py]: seriously need some `properties` stuff in here.

File: SimPEG/EM/FDEM/RxFDEM.py [hunk 14]
```diff
@@ -11,17 +12,27 @@ class BaseRx(SimPEG.Survey.BaseRx):
...
+        ), "Orientation {0!s} not known. Orientation must be in 'x', 'y', 'z'."
+        " Arbitrary orientations have not yet been implemented.".format(
+            orientation
+        )
+        assert(
+            component in ['real', 'imag']
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/RxFDEM.py]:
```python
component = properties.StringChoice(
    'Component of the fields object',
    choices=('real', 'imag')
)
```

File: SimPEG/Examples/EM_FDEM_Mu_Inversion.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+    mtrue = mur[active]
+
+    # Maps
+    actMap = Maps.InjectActiveCells(mesh, active, mur_air, nC=mesh.nCz)
+    surj1Dmap = Maps.SurjectVertical1D(mesh)
+    murMap = Maps.Mu_relative(mesh)
+
+    # Mapping
+    muMap =  murMap * surj1Dmap * actMap
```
Comment (rowanc1) [inline SimPEG/Examples/EM_FDEM_Mu_Inversion.py]: should test with a `mui` inversion as well?

Comment (rowanc1) [inline SimPEG/Examples/EM_FDEM_Mu_Inversion.py]: I suppose this is an example. Could have it as an option?

File: SimPEG/Maps.py [hunk 16]
```diff
@@ -519,6 +520,56 @@ def inverse(self, m):
         return np.exp(Utils.mkvc(m))
 
 
+class ChiMap(IdentityMap):
```
Comment (rowanc1) [inline SimPEG/Maps.py]: @fourndo are you aware that this is coming?

File: SimPEG/Maps.py [hunk 17]
```diff
@@ -519,6 +520,56 @@ def inverse(self, m):
...
+    def deriv(self, m, v=None):
+        if v is not None:
+            return mu_0 * v
+        return mu_0 * sp.eye(self.nP)
+
+    def inverse(self, m):
+        return m / mu_0 - 1
+
+
+class Mu_relative(IdentityMap):
```
Comment (rowanc1) [inline SimPEG/Maps.py]: cool map. We might want to change `Maps.py` into a folder?

`Maps.EM.Mu_relative` etc.

File: SimPEG/Mesh/CylMesh.py [hunk 18]
```diff
@@ -367,6 +367,52 @@ def aveF2CCV(self):
                                           'yet implemented')
         return self._aveF2CCV
 
+
+    def getInterpolationMat(self, loc, locType='CC', zerosOutside=False):
```
Comment (rowanc1) [inline SimPEG/Mesh/CylMesh.py]: we didn't have this?

File: SimPEG/Mesh/TensorMesh.py [hunk 19]
```diff
@@ -217,7 +229,7 @@ def isInside(self, pts, locType='N'):
             inside = inside & (pts[:,i] >= tensor.min()-TOL) & (pts[:,i] <= tensor.max()+TOL)
         return inside
 
-    def getInterpolationMat(self, loc, locType='CC', zerosOutside=False):
+    def _getInterpolationMat(self, loc, locType='CC', zerosOutside=False):
```
Comment (rowanc1) [inline SimPEG/Mesh/TensorMesh.py]: what is the rename here?

File: SimPEG/Mesh/CylMesh.py (hunk 18 above)
Comment (rowanc1) [inline SimPEG/Mesh/CylMesh.py]: I guess you answered it in your notes above. CCv needs specific treatment in the Cyl mesh.

File: SimPEG/EM/FDEM/FieldsFDEM.py (hunk 10 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/FieldsFDEM.py]: we don't currently have an eSecondary receiver implemented, but I guess there is no harm in allowing those fields to be retrieved.

File: SimPEG/EM/Base.py (hunk 7 above)
Comment (lheagy) [inline SimPEG/EM/Base.py]: right now, the mu inversion is only implemented in the frequency domain, so if you were to set up a time domain problem, the attribute currently does not exist. 
Perhaps though, getting a 'muMap' does not exsist error would be ok in the mean-time?

File: SimPEG/EM/Base.py (hunk 5 above)
Comment (lheagy) [inline SimPEG/EM/Base.py]: Thanks for catching this

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 14 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/RxFDEM.py]: I am still not sure if 'component' is the right word. I had a hard time explaining this - generally when referring to 'component' the first association is with the spatial component. eg. "we are measuring the x-component of the electric field". Perhaps `part in ["real", "imag"]` ?

File: SimPEG/Examples/EM_FDEM_Mu_Inversion.py (hunk 15 above)
Comment (lheagy) [inline SimPEG/Examples/EM_FDEM_Mu_Inversion.py]: the tests are testing all formulations, so we will hit examples where the mui map is used. The tests could be expanded to include setting both mu and mui maps. I think generally, one would invert for 'mu', thinking in terms of mu-inverse is pretty unnatural.

File: SimPEG/Mesh/CylMesh.py (hunk 18 above)
Comment (lheagy) [inline SimPEG/Mesh/CylMesh.py]: yep

File: SimPEG/Mesh/TensorMesh.py (hunk 19 above)
Comment (lheagy) [inline SimPEG/Mesh/TensorMesh.py]: This is so I can call it from the CylMesh Interpolation. The Edge, CC and Face interpolation from Tensor Mesh are all valid for Cyl

image

File: SimPEG/Examples/EM_FDEM_Mu_Inversion.py (hunk 15 above)
Comment (lheagy) [inline SimPEG/Examples/EM_FDEM_Mu_Inversion.py]: I have added mui inversions to the testing

File: SimPEG/EM/Base.py (hunk 1 above)
Comment (lheagy) [inline SimPEG/EM/Base.py]: good call, thanks!

File: SimPEG/EM/Base.py (hunk 9 above)
Comment (lheagy) [inline SimPEG/EM/Base.py]: yep, good call

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 14 above)
Comment (rowanc1) [inline SimPEG/EM/FDEM/RxFDEM.py]: This would also become a boolean `real_component = True`

File: SimPEG/Mesh/TensorMesh.py (hunk 19 above)
Comment (rowanc1) [inline SimPEG/Mesh/TensorMesh.py]: Could alternatively call `super` or `BaseTensorMesh.getInterpolationMat(self, ...)`  for future reference. I think this is fine.

File: SimPEG/EM/FDEM/ProblemFDEM.py (hunk 12 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: I think that is everything for the problem. For a forward simulation, we could also validate that - [ ] mesh size > skin depth of lowest freq
- [ ] smallest cells are < skin depth of highest freq
(perhaps not error on these but warn)
- [ ] check that src loc, rx locs are in the domain

File: SimPEG/EM/FDEM/ProblemFDEM.py (hunk 11 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: :)

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 13 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/RxFDEM.py]: yes we do. I think this should be addressed in a separate pr though?

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 14 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/RxFDEM.py]: it could be - I think it is odd to treat them unequally though. Keeping it as a string, I could see us allowing:
`["real", "in-phase"]` and `["imag", "imaginary", "quadrature", "out-of-phase" ]`

@dougoldenburg: any suggestions on what to label `real` and `imag` data (does "component" make sense, or do you have another suggestion?)

File: SimPEG/Mesh/TensorMesh.py (hunk 19 above)
Comment (lheagy) [inline SimPEG/Mesh/TensorMesh.py]: ah, I didn't know super could do that. clever!

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 13 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/RxFDEM.py]: I am currently working on implementing this on the em/ref/tdem-hj branch

**Comment by lheagy** [comment]
@rowanc1. It looks like the default value for mu is now not being set for the FDEM problem. Another issue currently is that I wanted to inject properties into Maps (by inheriting BaseSimPEG from Props), but since Props imports Maps, it was circular. Perhaps we move BaseSimPEG elsewhere? 

cc @fwkoch

**Comment by lheagy** [comment]
@rowanc1: what are your thoughts on bringing this in?

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 14 above)
Comment (rowanc1) [inline SimPEG/EM/FDEM/RxFDEM.py]: I like `part`. Measure the x-`component` and take the real `part`.

File: SimPEG/EM/Base.py [hunk 20]
```diff
@@ -132,6 +145,26 @@ def MfMuiI(self):
             self._MfMuiI = self.mesh.getFaceInnerProduct(self.mui, invMat=True)
```
Comment (rowanc1) [inline SimPEG/EM/Base.py]: Is this used anywhere?

File: SimPEG/EM/FDEM/ProblemFDEM.py [hunk 21]
```diff
@@ -15,28 +20,46 @@ class BaseFDEMProblem(BaseEMProblem):
...
+        \\\(\\\mathbf{h}\\\)
 
     """
 
     surveyPair = SurveyFDEM
     fieldsPair = FieldsFDEM
 
+    mu, muMap, muDeriv = Props.Invertible(
+        "Magnetic Permeability (H/m)",
+        default=mu_0
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: did the default thing get fixed?

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 14 above)
Comment (rowanc1) [inline SimPEG/EM/FDEM/RxFDEM.py]: See #465

File: SimPEG/EM/FDEM/ProblemFDEM.py (hunk 12 above)
Comment (rowanc1) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: I added a bunch of the checks/summary to PR #519 
Copied over the other comment to #464

File: SimPEG/EM/FDEM/RxFDEM.py [hunk 22]
```diff
@@ -11,17 +12,27 @@ class BaseRx(SimPEG.Survey.BaseRx):
...
+            ), "'component' must be 'real' or 'imag', not {0!s}".format(
+                component
+            )
 
         self.projComp = orientation
         self.component = component
 
-        SimPEG.Survey.BaseRx.__init__(self, locs, rxType=None) #TODO: remove rxType from baseRx
+        # TODO: remove rxType from baseRx
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/RxFDEM.py]: #524

File: SimPEG/EM/FDEM/SrcFDEM.py [hunk 23]
```diff
@@ -488,6 +488,33 @@ def s_e(self, prob):
 
             return -C.T * (MMui_s * self.bPrimary(prob))
 
+    def s_eDeriv(self, prob, v, adjoint=False):
+        if not hasattr(prob, 'muMap') or not hasattr(prob, 'muiMap'):
+            return Zero()
+        else:
+            formulation = prob._formulation
+
+            if formulation is 'EB':
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/SrcFDEM.py]: this should be changed to `==` rather than `is`

File: SimPEG/EM/FDEM/SrcFDEM.py [hunk 24]
```diff
@@ -488,6 +488,33 @@ def s_e(self, prob):
...
+                    self.bPrimary(prob)
+                ) * prob.muiDeriv
+                C = prob.mesh.edgeCurl
+
+                if adjoint:
+                    return -MMui_sDeriv.T * (C * v)
+
+                return -C.T * (MMui_sDeriv * v)
+
+            elif formulation is 'HJ':
```
Comment (rowanc1) [inline SimPEG/EM/FDEM/SrcFDEM.py]: `==`

File: SimPEG/Examples/EM_FDEM_Mu_Inversion.py [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+    )
+
+    srcList = [
+        FDEM.Src.MagDipole([bzr], freq, srcLoc, orientation='Z')
+        for freq in freqs
+    ]
+
+    surveyFD = FDEM.Survey(srcList)
+    prbFD = FDEM.Problem3D_b(
+        mesh, sigma=surj1Dmap * sigma, muMap=muMap, Solver=Solver
```
Comment (rowanc1) [inline SimPEG/Examples/EM_FDEM_Mu_Inversion.py]: This is cool that you use a map to set sigma, but don't put it in the model location.

File: SimPEG/EM/Base.py (hunk 20 above)
Comment (lheagy) [inline SimPEG/EM/Base.py]: not at the moment.

Comment (lheagy) [inline SimPEG/EM/Base.py]: it might be used in the TDEM implementation later on

File: SimPEG/EM/FDEM/RxFDEM.py (hunk 14 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/RxFDEM.py]: sounds good. we can address this in a separate pr

File: SimPEG/EM/FDEM/ProblemFDEM.py (hunk 21 above)
Comment (lheagy) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: yes, the default bug was fixed in #513

File: SimPEG/Examples/EM_FDEM_Mu_Inversion.py (hunk 25 above)
Comment (lheagy) [inline SimPEG/Examples/EM_FDEM_Mu_Inversion.py]: :)
